### DISCUSSION TITLE
Offline simulation

### DISCUSSION THREAD

**Discussion by xbh-sdju** [OpenRealTimeSimulation/SolverCodegen#9] [Q&A]
当我在 Microsoft Visual C++ 2010 Express 中编写脱机C++仿真测试文件时，它显示找不到我的 RLC_Circuit.hpp，但我已加载该文件。这真的很糟糕，RLC_Circuit.hpp是从CLI文件生成的，示例是操作手册中的RLC示例。希望得到你的帮助。非常感谢，附件是我的项目文件。
[rlc2.zip](https://github.com/OpenRealTimeSimulation/SolverCodegen/files/8563058/rlc2.zip)

**Comment by OpenRealTimeSimulation** [comment]
Hi @xbh-sdju,

If I translated correctly, in your testbench you are creating in Visual C++, RLC_Circuit.hpp cannot be found, despite being added to your project, correct?

When opening the rlc_tb.cpp file in Notepad++, I noticed the #include for the RLC_Circuit.hpp has a RS character in it, like seen here:

![image](https://user-images.githubusercontent.com/61946270/166808977-b90326ca-6e58-459e-8dac-6cf1f0969910.png)

This extra RS character might be causing the issue you are having.  These extra characters might occur when copying and pasting from different files with different text encoding (ANSI, UTF, etc.) and/or languages.

I noticed myself that copying code from our CLI User Guide PDF to plain-text also introduces other characters and unintended spaces in the code which shouldn't normally happen and may cause issues later down the road.  In later versions of the user guide, I'll see about fixing this issue.  In the meantime, I recommend manually copying or writing code from the user guide and to double check for unusual characters appearing in the code using your text editor.

Let me know if this answer helped you fix the problem.

**Reply by xbh-sdju** [reply]
Indeed, the issue has been resolved. But I have a new problem again, line 29 shows that I do not define RLC_Circuit and the double type is not allowed. Please take a look at my source file, thank you very much.
![rlc2](https://user-images.githubusercontent.com/84470481/167085575-af4cb612-9bce-47a1-9384-405e2c8abcc9.png)

**Comment by OpenRealTimeSimulation** [comment]
Hello @xbh-sdju ,

Sorry for the late reply.

> xbh-sdju:  Indeed, the issue has been resolved. But I have a new problem again, line 29 shows that I do not define RLC_Circuit and the double type is not allowed. Please take a look at my source file, thank you very much.

The solver function for RLC_Circuit model should be called `RLC_Circuit_solver`.  See the header file for the solver function you generated.  The solver function's name is always its model name with `_solver` appended to that.

Also make sure in your specific testbench for the `RLC_Circuit_solver()` call that `x_out` is an array of size 2 and `y_current_inductor1` is either a pointer to a real valued object (real := double) or an address to said object.  In your rlc2.zip project, rlc_tb.cpp source, `y_current_inductor1` is a double-type variable object, so pass its address to the solver function like so `&y_current_inductor1` to get solver function call like this:

`RLC_Circuit_solver<0,double>(x_out, &y_current_inductor1 );`

Always check signature of the generated solver function in the header file to make sure to know the function name and names/types/order of I/O arguments for said function.

The above should hopefully fix the errors you are receiving, including one about double type.  If the compiler still complains about double, check for other errors, and make sure the compiler is set for C++ file input and build.

Let me know if this answer helped you fix the problem.

**Reply by xbh-sdju** [reply]
The solver function name is changed to RLC_Circuit_solver no longer reports an error, but the double type name error is still displayed. The function template has been confirmed to be of double type, and we are waiting for your reply.
![image](https://user-images.githubusercontent.com/84470481/169828920-a3d4980f-bf4e-4f13-b318-6fe5551d3d9c.png)

**Comment by OpenRealTimeSimulation** [comment]
Hi @xbh-sdju,

Looking through your rlc2.zip file, it looks like in the  RLC_Circuit.hpp header you have the `template <>` part of the solver function commented out, as seen here:

![image](https://user-images.githubusercontent.com/61946270/170887212-ea889430-fe29-4857-bde8-1ac292ae8ec1.png)

Uncomment the template <> part of the function definition and remove the `typedef float real;` line from the solver function's header; you can put the real typedef elsewhere like in your testbench if you need it, but recommended to use double instead of float for offline testing for higher precision/range.  Restoring the template <> part should hopefully fix the error you are getting regarding double type.

The problem you had was the solver function was being called as an instanced function template with arguments 0 and double in the testbench code but by having the template <> commented out in the solver function's header, you made the function no longer a template, causing the compiler to throw an error.

Let me know how it goes with this fix.

**Comment by OpenRealTimeSimulation** [comment]
Discussion closed and locked due to inactivity.
